modules/examples/sources/threaded-diskq-source/threaded-diskq-source.c

```diff
@@ -0,0 +1,241 @@
+
+  if (!file)
+    {
+      msg_error("Error opening diskq file", evt_tag_str("file", self->filename));
+      return FALSE;
+    }
+
+  gchar file_signature[5] = {0};
+  if (fread(file_signature, 4, 1, file) == 0)
+    {
+      msg_error("Error reading diskq file signature", evt_tag_str("file", self->filename));
+      fclose(file);
+      return FALSE;
+    }
+
+  fclose(file);
+
+  if (strcmp(file_signature, "SLRQ") == 0)
+    self->diskq_options.reliable = TRUE;
+
+  if (self->diskq_options.reliable)
+    {
+      self->diskq_options.disk_buf_size = 128;
+      self->diskq_options.mem_buf_size = 1024 * 1024;
+      self->queue = log_queue_disk_reliable_new(&self->diskq_options, NULL);
+    }
+  else
+    {
+      self->diskq_options.disk_buf_size = 1;
+      self->diskq_options.mem_buf_size = 128;
+      self->diskq_options.qout_size = 128;
+      self->queue = log_queue_disk_non_reliable_new(&self->diskq_options, NULL);
+    }
+
+  if (!log_queue_disk_load_queue(self->queue, self->filename))
+    {
+      msg_error("Error loading diskq", evt_tag_str("file", self->filename));
+      return FALSE;
+    }
+
+  return TRUE;
+}
+
+static gboolean
+_new_diskq_file_exists(ThreadedDiskqSourceDriver *self, const struct stat *new_diskq_file_stat)
+{
+  return new_diskq_file_stat->st_mtime != self->diskq_file_stat.st_mtime
+         || new_diskq_file_stat->st_size != self->diskq_file_stat.st_size;
+}
+
+static gboolean
+_open_diskq(LogThreadedFetcherDriver *s)
+{
+  ThreadedDiskqSourceDriver *self = (ThreadedDiskqSourceDriver *) s;
+
+  struct stat new_diskq_file_stat;
+  if (stat(self->filename, &new_diskq_file_stat) != 0)
+    {
+      msg_verbose("Diskq file does now exist, retrying", evt_tag_str("file", self->filename));
+      return FALSE;
+    }
+
+  if (self->waiting_for_file_change)
+    {
+      if(!_new_diskq_file_exists(self, &new_diskq_file_stat))
+        {
+          msg_debug("Still waiting for new file", evt_tag_str("file", self->filename));
+          return FALSE;
+        }
+
+      self->waiting_for_file_change = FALSE;
+    }
+
+  if (!_load_queue(self))
+    return FALSE;
+
+  self->diskq_file_stat = new_diskq_file_stat;
+  return TRUE;
+}
+
+static void
+_close_diskq(LogThreadedFetcherDriver *s)
+{
+  ThreadedDiskqSourceDriver *self = (ThreadedDiskqSourceDriver *) s;
+
+  log_queue_unref(self->queue);
+  self->queue = NULL;
+}
+
+/* runs in a dedicated thread */
+static LogThreadedFetchResult
+_fetch(LogThreadedFetcherDriver *s)
+{
+  ThreadedDiskqSourceDriver *self = (ThreadedDiskqSourceDriver *) s;
+  LogPathOptions local_options = LOG_PATH_OPTIONS_INIT;
+
+  LogMessage *msg = log_queue_pop_head(self->queue, &local_options);
+
+  if (!msg)
+    {
+      msg_debug("Diskq file has been read, waiting for new file", evt_tag_str("file", self->filename));
+      _close_diskq(s);
+      self->waiting_for_file_change = TRUE;
+      return (LogThreadedFetchResult)
+      {
+        THREADED_FETCH_NOT_CONNECTED, NULL
+      };
```

Comment:
I now understand that it is not an artifact, I just didn't notice we are returning an instance of a struct. indentation could probably be improved.

Reply:
Yeah, it is ugly this way, but `astyle` forced me to do that.